### PR TITLE
LibJS: Consolidate sources of system time zone to one location in Date

### PR DESCRIPTION
**Pull request by trflynn89**
A "normative" ECMA-262 change followed by the editorial fallout in Temporal. Though the ECMA-262 change is normative, there's no behavior change.

### CODE REVIEW

**Comment by trflynn89** [comment]
Ref #15525

File: Userland/Libraries/LibJS/Runtime/Date.cpp [hunk 1]
```diff
@@ -285,21 +291,158 @@ double local_tza(double time, [[maybe_unused]] bool is_utc, Optional<StringView>
...
+        auto possible_instants = get_named_time_zone_epoch_nanoseconds(local_time_zone, year_from_time(time), month_from_time(time) + 1, date_from_time(time), hour_from_time(time), min_from_time(time), sec_from_time(time), ms_from_time(time), 0, 0);
+
+        // b. NOTE: The following steps ensure that when t represents local time repeating multiple times at a negative time zone transition (e.g. when the daylight saving time ends or the time zone offset is decreased due to a time zone rule change) or skipped local time at a positive time zone transition (e.g. when the daylight saving time starts or the time zone offset is increased due to a time zone rule change), t is interpreted using the time zone offset before the transition.
+        Crypto::SignedBigInteger disambiguated_instant;
+
+        // c. If possibleInstants is not empty, then
+        if (!possible_instants.is_empty()) {
+            //     i. Let disambiguatedInstant be possibleInstants[0].
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Date.cpp:420]:
```suggestion
            // i. Let disambiguatedInstant be possibleInstants[0].
```

**Review by linusg — COMMENTED** [review]
Great work! Only found one mistake in a comment, everything else looks good. I like the branch name :^)

File: Userland/Libraries/LibJS/Runtime/Date.cpp [hunk 2]
```diff
@@ -285,21 +291,158 @@ double local_tza(double time, [[maybe_unused]] bool is_utc, Optional<StringView>
...
+    static Crypto::SignedBigInteger const max_bigint { NumericLimits<i64>::max() };
+
+    // The provided epoch (nano)seconds value is potentially out of range for AK::Time and subsequently
+    // get_time_zone_offset(). We can safely assume that the TZDB has no useful information that far
+    // into the past and future anyway, so clamp it to the i64 range.
+    if (value < min_bigint)
+        return NumericLimits<i64>::min();
+    if (value > max_bigint)
+        return NumericLimits<i64>::max();
+    return value.to_base(10).to_int<i64>().value();
```
Comment (davidot) [inline Userland/Libraries/LibJS/Runtime/Date.cpp:328]: You could also just get the sign and use `SignedBigInteger::to_u64` so we don't have to go via a String.
(Could also make that a method on SignedBigInteger)

**Review by davidot — APPROVED** [review]
Looks good, just one small optional comment.
But feel free to ignore it, my brain seems to be in optimization mode

File: Userland/Libraries/LibJS/Runtime/Date.cpp (hunk 2 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Runtime/Date.cpp:328]: Or alternatively just leave a FIXME

Comment (trflynn89) [reply Userland/Libraries/LibJS/Runtime/Date.cpp:330]: Added a FIXME - I had the same thought, but this was copy-pasted from an existing Temporal AO, and I didn't wanna deal with unsigned-to-signed overflow here (which is UB, IIRC) :upside_down_face:. It seems odd that `SignedBigInteger` even has a `to_u64` method, so I think a follow-up to make that `to_i64` and use it here is in order.

**Comment by linusg** [comment]
This has conflicts now, sorry about that :upside_down_face:

**Review by linusg — APPROVED** [review]

**Pull request merged**